### PR TITLE
Encrypted files

### LINKED ISSUE #1862: Untrack frontend/model/contracts/manifests.json on `master`
### Problem

`frontend/model/contracts/manifests.json` is going to change regularly because contracts are signed now (and random keys are generated if the developer doesn't have an existing key).

### Solution

Untrack this file.

### PR DESCRIPTION
**Pull request by corrideat**
- Closes #1862
- Creates fileUpload and fileDownload APIs

### CODE REVIEW

File: frontend/model/database.js [hunk 1]
```diff
@@ -164,19 +164,70 @@ const files = localforage.createInstance({
...
 sbp('sbp/selectors/register', {
-  'gi.db/files/save': function (url: string, blob: Blob): Promise<*> {
-    return files.setItem(url, blob).then(v => {
-      console.log('successfully saved:', url)
-    }).catch(e => {
-      console.error('error saving:', url, e)
+  'gi.db/files/save': async function (cacheKey: string, blob: Blob): Promise<*> {
+    const keys = await files.getItem('keys') ?? []
+
+    return sbp('okTurtles.eventQueue/queueEvent', 'gi.db/files', () => {
```
Comment (taoeffect) [inline frontend/model/database.js:180]: can you add a comment explaining the reasoning for why an eventQueue is being used?

**Review by taoeffect — CHANGES_REQUESTED** [review]
Screenshot 2024-02-22 at 9 05 28 AM
Screenshot 2024-02-22 at 9 05 43 AM
Screenshot 2024-02-22 at 9 06 39 AM

**Review by taoeffect — CHANGES_REQUESTED** [review]
```
Running "exec:flow" (exec) task
Error -------------------------------------------------------------------------------- frontend/model/database.js:221:68

Cannot expect `Promise` as the return type of async function because `Blob` [1] is incompatible with implicitly-returned
undefined in type argument `R` [2]. [incompatible-return]

   frontend/model/database.js:221:68
    221|   'gi.db/files/delete': async function (cacheKey: string): Promise<Blob> {
                                                                            ^^^^ [1]

References:
   /private/tmp/flow/flowlib_e73d01e77da1cc86_501/core.js:1807:24
   1807| declare class Promise<+R> {
                                ^ [2]


Error ---------------------------------------------------------------------------- shared/domains/chelonia/files.js:7:24

Cannot resolve module `~/shared/multiformats/bytes.js`. [cannot-resolve-module]

   7| import { coerce } from '~/shared/multiformats/bytes.js'
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Review by taoeffect — CHANGES_REQUESTED** [review]
When voting on a poll:

Screenshot 2024-02-22 at 9 12 40 AM

File: frontend/controller/actions/identity.js [hunk 2]
```diff
@@ -495,6 +495,7 @@ export default (sbp('sbp/selectors/register', {
     } catch (e) {
       console.error(`${e.name} during logout: ${e.message}`, e)
     }
+    sbp('gi.db/files/clear')
```
Comment (taoeffect) [inline frontend/controller/actions/identity.js]: Can you add a comment here, something like, `// clear file cache`?

Comment (taoeffect) [reply frontend/controller/actions/identity.js]: Also, maybe it makes sense to make the selector say that

File: frontend/model/contracts/manifests.json [hunk 3]
```diff
@@ -1,7 +0,0 @@
-{
-  "manifests": {
-    "gi.contracts/chatroom": "z9brRu3VXcbQ8j8sZKgGtmBWrmhRPWYyYPCAjrwUFvteX93Hn8MV",
-    "gi.contracts/group": "z9brRu3VFeizdYVsxnbF4NfhqNedjVbrhNgd3Mzu9frebx1erEXB",
-    "gi.contracts/identity": "z9brRu3VU9rsSSW22TcEUGV3yKi9zXCfDcMGfGefgie2yUDszkpb"
-  }
-}
```
Comment (taoeffect) [inline frontend/model/contracts/manifests.json:7]: why was this deleted?

File: frontend/model/database.js [hunk 4]
```diff
@@ -164,19 +164,76 @@ const files = localforage.createInstance({
...
+
+    // We need to perform several operations in the DB, which includes
+    // the operation requested (i.e., saving a file) and updating the `keys`
+    // entry for caching, as well as possibly deleting cached entries.
+    // Because the `keys` entry is a point of contention among calls and we
+    // can only perform a single storage operation at the same time, we use
+    // a queue to ensure that all operations are done atomically.
+    return sbp('okTurtles.eventQueue/queueEvent', 'gi.db/files', () => {
+      // `_` prefix to avoid overlapping with the special `keys` key
+      return files.setItem('_' + cacheKey, blob).then(async v => {
```
Comment (taoeffect) [inline frontend/model/database.js]: Instead of doing it this way, how about reversing it to simplify things?

Use the `cacheKey` as-is, and have a special `__keys` key that cannot be set using `'gi.db/files/save'`.

File: frontend/model/database.js [hunk 5]
```diff
@@ -164,19 +164,76 @@ const files = localforage.createInstance({
...
+        await files.setItem('keys', keys)
+      }
+    }).catch(e => {
+      console.error('[gi.db/files/load] Error updating keys')
+    })
   },
-  'gi.db/files/delete': function (url: string): Promise<Blob> {
-    return files.removeItem(url)
+  'gi.db/files/clear': async function (): Promise<void> {
+    await archive.clear()
```
Comment (taoeffect) [inline frontend/model/database.js]: did you mean `files.clear()`?

File: frontend/model/database.js [hunk 6]
```diff
@@ -164,19 +164,76 @@ const files = localforage.createInstance({
...
 
+const maxFileEntries = 100
+
 sbp('sbp/selectors/register', {
-  'gi.db/files/save': function (url: string, blob: Blob): Promise<*> {
-    return files.setItem(url, blob).then(v => {
-      console.log('successfully saved:', url)
-    }).catch(e => {
-      console.error('error saving:', url, e)
+  'gi.db/files/save': async function (cacheKey: string, blob: Blob): Promise<*> {
```
Comment (taoeffect) [inline frontend/model/database.js]: If this is being used as a cache, then maybe we should rename it to `gi.db/filesCache`

File: frontend/model/database.js (hunk 4 above)
Comment (corrideat) [reply frontend/model/database.js]: It was based on this: <https://en.wikipedia.org/wiki/Hilbert%27s_paradox_of_the_Grand_Hotel>

I could do it the other way, but the special key still needs to be checked for and, throw an error, maybe?

File: frontend/controller/actions/identity.js (hunk 2 above)
Comment (corrideat) [reply frontend/controller/actions/identity.js]: The name is to keep it consistent with the rest of the `gi.db/` selectors

File: frontend/views/components/Avatar.vue [hunk 7]
```diff
@@ -31,70 +31,66 @@ export default ({
...
-    if (this.blobURL && !this.objectURL) {
-      // trigger the watcher
-      this.setBlobURL(this.blobURL)
+    if (typeof this.src === 'object') {
+      this.downloadFile(this.src).catch((e) => {
+        console.error('[Avatar.vue] Error in downloadFile', e)
+      })
     }
   },
+  beforeUnmount () {
```
Comment (taoeffect) [inline frontend/views/components/Avatar.vue]: `beforeUnmount` is not a thing

File: frontend/model/database.js (hunk 4 above)
Comment (taoeffect) [reply frontend/model/database.js]: > I could do it the other way, but the special key still needs to be checked for and, throw an error, maybe?

Yeah 👍

File: shared/domains/chelonia/files.js [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+        resolve(length)
+      },
+      abort (reason) {
+        reject(reason)
+      }
+    }))
+  })
+  const cidPromise = createCIDfromStream(cidStream)
+  return Promise.all([lengthPromise, cidPromise])
+}
```
Comment (taoeffect) [inline shared/domains/chelonia/files.js:83]: rly cool 👍

File: shared/domains/chelonia/files.js [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+      } catch (e) {
+        controller.error(e)
+      }
+    }
+  })
+}
+
+export const aes256gcmHandlers: any = {
+  upload: (chelonia: Object, manifestOptions: Object) => {
+    let IKM = manifestOptions['cipher-params']?.IKM
```
Comment (taoeffect) [inline shared/domains/chelonia/files.js:147]: Can you add a comment explaining what "IKM" stands for?

**Review by taoeffect — DISMISSED** [review]
LGTM, once the tests pass

**Review by taoeffect — APPROVED** [review]
Fantastic work @corrideat!!! 😄  🎉 🥇
